artiq/test/test_scheduler.py

```diff
@@ -82,6 +60,65 @@ def get(self):
         self._next_rid += 1
         return rid
 
+class SchedulerMonitor:
+    def __init__(self, test):
+        self.test = test
+        self.experiments = {}
+        self.last_status = {}
+        self.flags = {"arrive": {}, "leave": {}}
+        self.flow_map = {   # current status -> possible move
+            "": {"pending"},
+            "pending": {"preparing", "flushing", "deleting"},
+            "preparing": {"prepare_done", "deleting"},
+            "prepare_done": {"running", "deleting"},
+            "running": {"run_done", "paused", "deleting"},
+            "run_done": {"analyzing", "deleting"},
+            "analyzing": {"deleting"},
+            "deleting": {},
+            "paused": {"running"},
+            "flushing": {"preparing"}
+        }
+
+    def record(self, mod):
+        process_mod(self.experiments, mod)
+        for rid, exp_info in self.experiments.items():
+            if rid not in self.last_status.keys():
+                self.last_status[rid] = ""
+            if exp_info["status"] != self.last_status[rid]:
+                self.test.assertIn(exp_info["status"], self.flow_map[self.last_status[rid]])
+
+                if rid in self.flags["arrive"].keys():
+                    if exp_info["status"] in self.flags["arrive"][rid].keys():
+                        self.flags["arrive"][rid][exp_info["status"]].set()
+                if rid in self.flags["leave"].keys():
+                    if self.last_status[rid] in self.flags["leave"][rid].keys():
+                        self.flags["leave"][rid][self.last_status[rid]].set()
+
+                self.last_status[rid] = exp_info["status"]
+                return
+
+    async def wait_until(self, rid, condition, status):
+        # condition : "arrive", "leave"
```

Comment:
Use ``assert``?

Reply:
I don't think it is needed here as this condition might not be the things we need to check.

Reply:
Then this comment is confusing.

Reply:
Which part is confusing?

> might not be the things we need to check

I mean "not for assert". 
`wait_until` is like an indicator that some stage has been reached so the condition of other rid can be asserted.

Reply:
![Screenshot from 2022-08-04 10-52-39](https://user-images.githubusercontent.com/48083317/182752746-9b6b1ed9-5bcd-4cff-8a8a-b228435d1d70.png)
In the case like test_pending_priority, one of the rid meant not to be finished. So test for assertion is after `wait_until`.
Also line 88 would check whether the exp go through the stage in correct order